### PR TITLE
[RFC] Pooling and LeakyReLu

### PR DESCRIPTION
**Pull request by vchuravy**
This consolidates three important operators needed for floating point support.

One general question is whether we should enable operators for all DTypes or only for floating point. Compare the implementation of BatchNorm to LeakyReLu for that.

This depends on https://github.com/dmlc/mshadow/pull/134

@Godricly Thank you for BatchNorm. I merged my changes into the commit, but kept the main attribution for you.

### CODE REVIEW

File: src/operator/batch_norm.cc [hunk 1]
```diff
@@ -10,12 +10,31 @@
 namespace mxnet {
 namespace op {
 template<>
-Operator *CreateOp<cpu>(BatchNormParam param) {
-  return new BatchNormOp<cpu>(param);
+Operator *CreateOp<cpu>(BatchNormParam param, int dtype) {
```
Comment (vchuravy) [inline src/operator/batch_norm.cc]: Should this be enabled for the integral dtypes?

File: src/operator/cudnn_batch_norm-inl.h [hunk 2]
```diff
@@ -156,22 +157,22 @@ class CuDNNBatchNormOp : public Operator {
...
+      in_data[cudnnbatchnorm::kGamma].get_with_shape<gpu, 1, DType>(Shape1(shape_[1]), s);
+    Tensor<gpu, 1, DType> dbeta =
+      in_grad[cudnnbatchnorm::kBeta].get_with_shape<gpu, 1, DType>(Shape1(shape_[1]), s);
+    Tensor<gpu, 1, DType> dgamma =
+      in_grad[cudnnbatchnorm::kGamma].get_with_shape<gpu, 1, DType>(Shape1(shape_[1]), s);
+    Tensor<gpu, 1, DType> save_mean =
+      out_data[cudnnbatchnorm::kMean].get_with_shape<gpu, 1, DType>(Shape1(shape_[1]), s);
+    Tensor<gpu, 1, DType> save_inv_var =
+      out_data[cudnnbatchnorm::kInvVar].get_with_shape<gpu, 1, DType>(Shape1(shape_[1]), s);
+    DType a = static_cast<DType>(1.0f);
```
Comment (vchuravy) [inline src/operator/cudnn_batch_norm-inl.h]: @Godricly Does cudnn really expects the parameters to be of the correct DTypes?

It should probably be similar to https://github.com/dmlc/mxnet/blob/8e88cb7148fb50c1191b64b2bd3f2d33de023c05/src/operator/cudnn_convolution-inl.h#L61

**Comment by Godricly** [comment]
The testing is not added yet. I tried on cudnnv5. But I'm not sure about cudnnv4 or self implemented batch norm.

File: src/operator/cudnn_batch_norm-inl.h (hunk 2 above)
Comment (Godricly) [inline src/operator/cudnn_batch_norm-inl.h]: Is that static_cast I used you worried about?

File: src/operator/batch_norm-inl.h [hunk 3]
```diff
@@ -71,49 +71,51 @@ class BatchNormOp : public Operator {
     }
 
     Stream<xpu> *s = ctx.get_stream<xpu>();
-    const real_t scale = static_cast<real_t>(in_data[batchnorm::kData].shape_[1]) /
-                         static_cast<real_t>(in_data[batchnorm::kData].shape_.Size());
-    Tensor<xpu, 4> data;
-    Tensor<xpu, 4> out;
+    const ScalarExp<DType> scale = scalar<DType>(
+        in_data[batchnorm::kData].shape_[1] / in_data[batchnorm::kData].shape_.Size());
```
Comment (Godricly) [inline src/operator/batch_norm-inl.h]: I think we should keep the computation in real_t. The shape are int like type.

**Comment by vchuravy** [comment]
Any ideas about this?

```
In file included from /home/wallnuss/src/mxnet/mshadow/mshadow/./extension.h:10:0,
                 from /home/wallnuss/src/mxnet/mshadow/mshadow/tensor.h:814,
                 from include/mxnet/./base.h:13,
                 from include/mxnet/operator.h:18,
                 from src/operator/./batch_norm-inl.h:12,
                 from src/operator/batch_norm.cc:8:
/home/wallnuss/src/mxnet/mshadow/mshadow/././expr_engine-inl.h:382:22: warning: inline function ‘static void mshadow::expr::ExpComplexEngine<SV, RV, E, DType>::Eval(RV*, const E&) [with SV = mshadow::sv::saveto; RV = mshadow::Tensor<mshadow::cpu, 1, mshadow::half::half_t>; E = mshadow::expr::BinaryMapExp<mshadow::op::mul, mshadow::expr::ScalarExp<mshadow::half::half_t>, mshadow::expr::ReduceTo1DExp<mshadow::Tensor<mshadow::cpu, 4, mshadow::half::half_t>, mshadow::half::half_t, mshadow::red::sum, 3>, mshadow::half::half_t, 7>; DType = mshadow::half::half_t]’ used but never defined
   inline static void Eval(RV *dst, const E &exp);
                      ^~~~
/home/wallnuss/src/mxnet/mshadow/mshadow/././expr_engine-inl.h:382:22: warning: inline function ‘static void mshadow::expr::ExpComplexEngine<SV, RV, E, DType>::Eval(RV*, const E&) [with SV = mshadow::sv::saveto; RV = mshadow::Tensor<mshadow::cpu, 1, mshadow::half::half_t>; E = mshadow::expr::BinaryMapExp<mshadow::op::mul, mshadow::expr::ScalarExp<mshadow::half::half_t>, mshadow::expr::ReduceTo1DExp<mshadow::expr::UnaryMapExp<mxnet::op::mshadow_op::square, mshadow::expr::BinaryMapExp<mshadow::op::minus, mshadow::Tensor<mshadow::cpu, 4, mshadow::half::half_t>, mshadow::expr::MakeTensorExp<mshadow::expr::Broadcast1DExp<mshadow::Tensor<mshadow::cpu, 1, mshadow::half::half_t>, mshadow::half::half_t, 4, 3>, mshadow::Tensor<mshadow::cpu, 1, mshadow::half::half_t>, 4, mshadow::half::half_t>, mshadow::half::half_t, 3>, mshadow::half::half_t, 3>, mshadow::half::half_t, mshadow::red::sum, 3>, mshadow::half::half_t, 7>; DType = mshadow::half::half_t]’ used but never defined
```

File: src/operator/cudnn_batch_norm-inl.h (hunk 2 above)
Comment (vchuravy) [inline src/operator/cudnn_batch_norm-inl.h]: Yes, I have to look at the call cudnn expects.

File: src/operator/batch_norm-inl.h (hunk 3 above)
Comment (vchuravy) [inline src/operator/batch_norm-inl.h]: Yes you are right. I will address that.

**Comment by vchuravy** [comment]
@tqchen @piiswrong So right now we have to wrap all parameters to be of `DType`. The general solution (which I am looking into right now) is to enable mshadow expressions with mixed DTypes (primarily scalars of type `real_t` and tensors of `DType`).

File: src/operator/cudnn_batch_norm-inl.h (hunk 2 above)
Comment (Godricly) [inline src/operator/cudnn_batch_norm-inl.h]: I think we don't need to use static_cast according to section 2.6 in cudnn doc. You can double check that.

Comment (Godricly) [inline src/operator/cudnn_batch_norm-inl.h]: I just realized that the address is what passed in. So we need some switch case here to convert them into float or double.

**Comment by Godricly** [comment]
We have some problem here in infertype as bn uses aux_array. Working on a fix.

File: src/operator/cudnn_batch_norm-inl.h [hunk 4]
```diff
@@ -259,6 +260,26 @@ class CuDNNBatchNormProp : public OperatorProperty {
...
+      if ((*in_type)[i] == -1) {
+        (*in_type)[i] = dtype;
+      } else {
+        CHECK_EQ((*in_type)[i], dtype) << "This layer requires uniform type. "
+                                       << "Expected " << dtype << " v.s. given "
+                                       << (*in_type)[i] << " at " << ListArguments()[i];
+      }
+    }
+    out_type->clear();
+    out_type->push_back(dtype);
```
Comment (Godricly) [inline src/operator/cudnn_batch_norm-inl.h]: modify InferType so the length of each vector matches InferShape

**Comment by Godricly** [comment]
Did you merge changes from master to this branch? My previous builds in embedding could not pass  due to mshadow dependence. If was not able to compile until #2511 updated mshadow dependence.

**Comment by vchuravy** [comment]
Depends on https://github.com/dmlc/mshadow/pull/141

**Comment by Godricly** [comment]
maybe you should separate this pull into 2 and take bn part out first. The bn is functional on gpu with mshadow updates, but not for cpu side. You can get pooling and leak relu merged first. I can get fp64 working now with pooling and dropout. :laughing:

**Comment by vchuravy** [comment]
Split the PR

File: src/operator/cudnn_pooling-inl.h [hunk 5]
```diff
@@ -90,10 +91,10 @@ class CuDNNPoolingOp : public Operator {
...
-    Tensor<gpu, 4> m_out_grad = out_grad[pool_enum::kOut].get<gpu, 4, real_t>(s);
-    Tensor<gpu, 4> m_in_data = in_data[pool_enum::kData].get<gpu, 4, real_t>(s);
-    Tensor<gpu, 4> m_out_data = out_data[pool_enum::kOut].get<gpu, 4, real_t>(s);
-    Tensor<gpu, 4> m_in_grad = in_grad[pool_enum::kData].get<gpu, 4, real_t>(s);
+    Tensor<gpu, 4, DType> m_out_grad = out_grad[pool_enum::kOut].get<gpu, 4, DType>(s);
+    Tensor<gpu, 4, DType> m_in_data = in_data[pool_enum::kData].get<gpu, 4, DType>(s);
+    Tensor<gpu, 4, DType> m_out_data = out_data[pool_enum::kOut].get<gpu, 4, DType>(s);
+    Tensor<gpu, 4, DType> m_in_grad = in_grad[pool_enum::kData].get<gpu, 4, DType>(s);
     CHECK_EQ(s->dnn_handle_ownership_, mshadow::Stream<gpu>::OwnHandle);
     float alpha = 1.0f;
```
Comment (Godricly) [inline src/operator/cudnn_pooling-inl.h]: There two parameters should be typed as:
 ```c++
typename DataType<DType>::ScaleType
```
Double type pooling goes wrong with current type.

**Comment by vchuravy** [comment]
Will update this to only contain LeakyReLU 

@Godricly thank you for picking up the work in #2824

**Comment by piiswrong** [comment]
any updates?

**Comment by piiswrong** [comment]
updates?

**Comment by piiswrong** [comment]
Closing for now
